// is-type-of/dist/esm/types/std.js
// Javascript Standard Object
import { types, debuglog } from 'node:util';
import { hasOwnProperty, isInstanceOf } from './util.js';
const debug = debuglog('is-type-of:std');
// Array
/**
 * returns true if val is array, it won't check items of array
 */
export function isArray(val) {
    return Array.isArray(val);
}
// Function
/**
 * returns true if val is function
 */
export function isFunction(val) {
    return typeof val === 'function';
}
/**
 * returns true if val is generator function
 */
export function isGeneratorFunction(val) {
    return types.isGeneratorFunction(val);
}
/**
 * returns true if val is async function
 */
export function isAsyncFunction(val) {
    return types.isAsyncFunction(val);
}
/**
 * returns true if val is async generator function
 */
export function isAsyncGeneratorFunction(val) {
    if (!isFunction(val))
        return false;
    return val.constructor && val.constructor.name === 'AsyncGeneratorFunction';
}
// Object
/**
 * returns true if val is object
 */
export function isObject(val) {
    return typeof val === 'object' && val !== null;
}
/**
 * returns true if val is class
 * *Note:* "class" is supported in ECMAScript 6, and if the code is using some compiler or transpiler, the checking might fail
 */
export function isClass(val) {
    if (!isFunction(val))
        return false;
    const fnStr = Function.prototype.toString.call(val);
    if (/^class[\s{]/.test(fnStr))
        return true;
    debug('[isClass] fnStr: %s', fnStr);
    const fnBody = fnStr.replace(/^[^{]*{\s*/, '').replace(/\s*}[^}]*$/, '');
    debug('[isClass] fnBody: %s', fnBody);
    return (/classCallCheck\(/.test(fnBody) || /TypeError\("Cannot call a class as a function"\)/.test(fnBody));
}
/**
 * returns true if val is regular expression
 */
export function isRegExp(val) {
    return isInstanceOf(val, RegExp);
}
/**
 * returns true if val is instance of Date
 */
export function isDate(val) {
    return isInstanceOf(val, Date);
}
/**
 * returns true if val is instance of Error
 */
export function isError(val) {
    return isInstanceOf(val, Error);
}
/**
 * returns true if val is generator
 */
export function isGenerator(val) {
    return types.isGeneratorObject(val);
}
/**
 * returns true if val is promise
 */
export function isPromise(val) {
    return types.isPromise(val);
}
/**
 * returns true if val is like promise, if the object has `then` property, the checking will pass.
 */
export function isPromiseLike(val) {
    if (isPromise(val))
        return true;
    if (!isObject(val))
        return false;
    return hasOwnProperty(val, 'then') && typeof val.then === 'function';
}
