// onelogger/dist/commonjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setCustomLogger = setCustomLogger;
exports.setLogger = setLogger;
exports.setCoreLogger = setCoreLogger;
exports.getCustomLogger = getCustomLogger;
exports.getLogger = getLogger;
exports.getCoreLogger = getCoreLogger;
const Logger_js_1 = require("./Logger.js");
__exportStar(require("./Logger.js"), exports);
function setCustomLogger(loggerName, realLogger) {
    Logger_js_1.Logger.setRealLogger(loggerName, realLogger);
}
function setLogger(realLogger) {
    setCustomLogger('logger', realLogger);
}
function setCoreLogger(realLogger) {
    setCustomLogger('coreLogger', realLogger);
}
const loggers = new Map();
function getCustomLogger(loggerName, prefix) {
    const key = `${loggerName}-${prefix ?? ''}`;
    let logger = loggers.get(key);
    if (!logger) {
        logger = new Logger_js_1.Logger({ loggerName, prefix });
        loggers.set(key, logger);
    }
    return logger;
}
function getLogger(prefix) {
    return getCustomLogger('logger', prefix);
}
function getCoreLogger(prefix) {
    return getCustomLogger('coreLogger', prefix);
}
