// detect-port/dist/commonjs/wait-port.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WaitPortRetryError = void 0;
exports.waitPort = waitPort;
const node_util_1 = require("node:util");
const detect_port_js_1 = require("./detect-port.js");
const debug = (0, node_util_1.debuglog)('detect-port:wait-port');
function sleep(ms) {
    return new Promise(resolve => {
        setTimeout(resolve, ms);
    });
}
class WaitPortRetryError extends Error {
    retries;
    count;
    constructor(message, retries, count, options) {
        super(message, options);
        this.name = this.constructor.name;
        this.retries = retries;
        this.count = count;
        Error.captureStackTrace(this, this.constructor);
    }
}
exports.WaitPortRetryError = WaitPortRetryError;
async function waitPort(port, options = {}) {
    const { retryInterval = 1000, retries = Infinity } = options;
    let count = 1;
    async function loop() {
        debug('wait port %d, retries %d, count %d', port, retries, count);
        if (count > retries) {
            const err = new WaitPortRetryError('retries exceeded', retries, count);
            throw err;
        }
        count++;
        const freePort = await (0, detect_port_js_1.detectPort)(port);
        if (freePort === port) {
            await sleep(retryInterval);
            return loop();
        }
        return true;
    }
    return await loop();
}
