// @eggjs/utils/dist/commonjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EggType = exports.getFrameworkOrEggPath = exports.getLoadUnits = exports.getConfig = exports.getPlugins = exports.getFrameworkPath = void 0;
exports.detectType = detectType;
const node_path_1 = __importDefault(require("node:path"));
const promises_1 = __importDefault(require("node:fs/promises"));
const framework_js_1 = require("./framework.js");
const plugin_js_1 = require("./plugin.js");
const deprecated_js_1 = require("./deprecated.js");
// support import { getFrameworkPath } from '@eggjs/utils'
var framework_js_2 = require("./framework.js");
Object.defineProperty(exports, "getFrameworkPath", { enumerable: true, get: function () { return framework_js_2.getFrameworkPath; } });
var plugin_js_2 = require("./plugin.js");
Object.defineProperty(exports, "getPlugins", { enumerable: true, get: function () { return plugin_js_2.getPlugins; } });
Object.defineProperty(exports, "getConfig", { enumerable: true, get: function () { return plugin_js_2.getConfig; } });
Object.defineProperty(exports, "getLoadUnits", { enumerable: true, get: function () { return plugin_js_2.getLoadUnits; } });
var deprecated_js_2 = require("./deprecated.js");
Object.defineProperty(exports, "getFrameworkOrEggPath", { enumerable: true, get: function () { return deprecated_js_2.getFrameworkOrEggPath; } });
__exportStar(require("./import.js"), exports);
__exportStar(require("./error/index.js"), exports);
// support import utils from '@eggjs/utils'
exports.default = {
    getFrameworkPath: framework_js_1.getFrameworkPath,
    getPlugins: plugin_js_1.getPlugins, getConfig: plugin_js_1.getConfig, getLoadUnits: plugin_js_1.getLoadUnits,
    getFrameworkOrEggPath: deprecated_js_1.getFrameworkOrEggPath,
};
var EggType;
(function (EggType) {
    EggType["framework"] = "framework";
    EggType["plugin"] = "plugin";
    EggType["application"] = "application";
    EggType["unknown"] = "unknown";
})(EggType || (exports.EggType = EggType = {}));
/**
 * Detect the type of egg project
 */
async function detectType(baseDir) {
    const pkgFile = node_path_1.default.join(baseDir, 'package.json');
    let pkg;
    try {
        await promises_1.default.access(pkgFile);
    }
    catch {
        return EggType.unknown;
    }
    try {
        pkg = JSON.parse(await promises_1.default.readFile(pkgFile, 'utf-8'));
    }
    catch {
        return EggType.unknown;
    }
    if (pkg.egg?.framework) {
        return EggType.framework;
    }
    if (pkg.eggPlugin?.name) {
        return EggType.plugin;
    }
    return EggType.application;
}
