// is-type-of/dist/commonjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Long = exports.duplex = exports.writable = exports.readable = exports.stream = exports.buffer = exports.promise = exports.generator = exports.regexp = exports.regExp = exports.error = exports.date = exports.object = exports.array = exports.class = exports.asyncFunction = exports.generatorFunction = exports.function = exports.primitive = exports.NaN = exports.finite = exports.double = exports.long = exports.int32 = exports.int = exports.bigint = exports.bigInt = exports.nullOrUndefined = exports.null = exports.undefined = exports.symbol = exports.string = exports.number = exports.boolean = exports.isInstanceOf = void 0;
const is_js_1 = require("./is.js");
const primitive_js_1 = require("./types/primitive.js");
Object.defineProperty(exports, "null", { enumerable: true, get: function () { return primitive_js_1.isNull; } });
Object.defineProperty(exports, "undefined", { enumerable: true, get: function () { return primitive_js_1.isUndefined; } });
Object.defineProperty(exports, "NaN", { enumerable: true, get: function () { return primitive_js_1.isNaN; } });
const std_js_1 = require("./types/std.js");
Object.defineProperty(exports, "class", { enumerable: true, get: function () { return std_js_1.isClass; } });
Object.defineProperty(exports, "function", { enumerable: true, get: function () { return std_js_1.isFunction; } });
__exportStar(require("./types/primitive.js"), exports);
__exportStar(require("./types/std.js"), exports);
__exportStar(require("./types/node.js"), exports);
__exportStar(require("./types/external.js"), exports);
var util_js_1 = require("./types/util.js");
Object.defineProperty(exports, "isInstanceOf", { enumerable: true, get: function () { return util_js_1.isInstanceOf; } });
exports.default = is_js_1.is;
// Compatible API, recommand to use `is`
/**
 * @deprecated
 * @see {is.boolean}
 */
exports.boolean = is_js_1.is.boolean;
/**
 * @deprecated
 * @see {is.number}
 */
exports.number = is_js_1.is.number;
/**
 * @deprecated
 * @see {is.string}
 */
exports.string = is_js_1.is.string;
/**
 * @deprecated
 * @see {is.symbol}
 */
exports.symbol = is_js_1.is.symbol;
/**
 * @deprecated
 * @see {is.nullable}
 */
exports.nullOrUndefined = is_js_1.is.nullable;
/**
 * @deprecated
 * @see {is.bigInt}
 */
exports.bigInt = is_js_1.is.bigInt;
/**
 * @deprecated
 * @see {is.bigInt}
 */
exports.bigint = is_js_1.is.bigInt;
/**
 * @deprecated
 * @see {is.integer}
 */
exports.int = is_js_1.is.integer;
/**
 * @deprecated
 * @see {is.integer32}
 */
exports.int32 = is_js_1.is.integer32;
/**
 * @deprecated
 * @see {is.long}
 */
exports.long = is_js_1.is.long;
/**
 * @deprecated
 * @see {is.double}
 */
exports.double = is_js_1.is.double;
/**
 * @deprecated
 * @see {is.finite}
 */
exports.finite = is_js_1.is.finite;
/**
 * @deprecated
 * @see {is.primitive}
 */
exports.primitive = is_js_1.is.primitive;
/**
 * @deprecated
 * @see {is.generatorFunction}
 */
exports.generatorFunction = is_js_1.is.generatorFunction;
/**
 * @deprecated
 * @see {is.asyncFunction}
 */
exports.asyncFunction = is_js_1.is.asyncFunction;
/**
 * @deprecated
 * @see {is.array}
 */
exports.array = is_js_1.is.array;
/**
 * @deprecated
 * @see {is.object}
 */
exports.object = is_js_1.is.object;
/**
 * @deprecated
 * @see {is.date}
 */
exports.date = is_js_1.is.date;
/**
 * @deprecated
 * @see {is.error}
 */
exports.error = is_js_1.is.error;
/**
 * @deprecated
 * @see {is.regExp}
 */
exports.regExp = is_js_1.is.regExp;
/**
 * @deprecated
 * @see {is.regExp}
 */
exports.regexp = is_js_1.is.regExp;
/**
 * @deprecated
 * @see {is.generator}
 */
exports.generator = is_js_1.is.generator;
/**
 * @deprecated
 * @see {is.promiseLike}
 */
exports.promise = is_js_1.is.promiseLike;
/**
 * @deprecated
 * @see {is.buffer}
 */
exports.buffer = is_js_1.is.buffer;
/**
 * @deprecated
 * @see {is.stream}
 */
exports.stream = is_js_1.is.stream;
/**
 * @deprecated
 * @see {is.readable}
 */
exports.readable = is_js_1.is.readable;
/**
 * @deprecated
 * @see {is.writable}
 */
exports.writable = is_js_1.is.writable;
/**
 * @deprecated
 * @see {is.duplex}
 */
exports.duplex = is_js_1.is.duplex;
/**
 * @deprecated
 * @see {is.longObject}
 */
exports.Long = is_js_1.is.longObject;
