// is-type-of/dist/commonjs/is.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.is = void 0;
const primitive_js_1 = require("./types/primitive.js");
const std_js_1 = require("./types/std.js");
const node_js_1 = require("./types/node.js");
const external_js_1 = require("./types/external.js");
exports.is = {
    // Primitive
    /**
     * @see {isBoolean}
     */
    boolean: primitive_js_1.isBoolean,
    /**
     * @see {isNumber}
     */
    number: primitive_js_1.isNumber,
    /**
     * @see {isString}
     */
    string: primitive_js_1.isString,
    /**
     * @see {isSymbol}
     */
    symbol: primitive_js_1.isSymbol,
    /**
     * @see {isUndefined}
     */
    undefined: primitive_js_1.isUndefined,
    /**
     * @see {isNull}
     */
    null: primitive_js_1.isNull,
    /**
     * @see {isNullable}
     */
    nullable: primitive_js_1.isNullable,
    /**
     * @see {isBigInt}
     */
    bigInt: primitive_js_1.isBigInt,
    /**
     * @see {isPrimitive}
     */
    primitive: primitive_js_1.isPrimitive,
    /**
     * @see {isInteger}
     */
    integer: primitive_js_1.isInteger,
    /**
     * @see {isInteger32}
     */
    integer32: primitive_js_1.isInteger32,
    /**
     * @see {isLong}
     */
    long: primitive_js_1.isLong,
    /**
     * @see {isDouble}
     */
    double: primitive_js_1.isDouble,
    /**
     * @see {isFinite}
     */
    finite: primitive_js_1.isFinite,
    /**
     * @see {isNaN}
     */
    NaN: primitive_js_1.isNaN,
    /**
     * @see {isSafeInteger}
     */
    safeInteger: primitive_js_1.isSafeInteger,
    // Standard Object
    /**
     * @see {isFunction}
     */
    function: std_js_1.isFunction,
    /**
     * @see {isGeneratorFunction}
     */
    generatorFunction: std_js_1.isGeneratorFunction,
    /**
     * @see {isAsyncFunction}
     */
    asyncFunction: std_js_1.isAsyncFunction,
    /**
     * @see {isAsyncGeneratorFunction}
     */
    asyncGeneratorFunction: std_js_1.isAsyncGeneratorFunction,
    /**
     * @see {isClass}
     */
    class: std_js_1.isClass,
    /**
     * @see {isArray}
     */
    array: std_js_1.isArray,
    /**
     * @see {isObject}
     */
    object: std_js_1.isObject,
    /**
     * @see {isDate}
     */
    date: std_js_1.isDate,
    /**
     * @see {isError}
     */
    error: std_js_1.isError,
    /**
     * @see {isRegExp}
     */
    regExp: std_js_1.isRegExp,
    /**
     * @see {isGenerator}
     */
    generator: std_js_1.isGenerator,
    /**
     * @see {isPromise}
     */
    promise: std_js_1.isPromise,
    /**
     * @see {isPromiseLike}
     */
    promiseLike: std_js_1.isPromiseLike,
    // Node Object
    /**
     * @see {isBuffer}
     */
    buffer: node_js_1.isBuffer,
    /**
     * @see {isStream}
     */
    stream: node_js_1.isStream,
    /**
     * @see {isReadable}
     */
    readable: node_js_1.isReadable,
    /**
     * @see {isWritable}
     */
    writable: node_js_1.isWritable,
    /**
     * @see {isDuplex}
     */
    duplex: node_js_1.isDuplex,
    // External Object
    /**
     * @see {isLongObject}
     */
    longObject: external_js_1.isLongObject,
};
