// runscript/dist/commonjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runscript = exports.RunScriptTimeoutError = exports.RunScriptError = void 0;
exports.runScript = runScript;
const node_util_1 = require("node:util");
const node_assert_1 = __importDefault(require("node:assert"));
const node_path_1 = __importDefault(require("node:path"));
const node_child_process_1 = require("node:child_process");
const is_type_of_1 = require("is-type-of");
const debug = (0, node_util_1.debuglog)('runscript');
function isCmd() {
    if (process.platform !== 'win32') {
        return false;
    }
    try {
        const result = (0, node_child_process_1.spawnSync)('ls', {
            stdio: 'pipe',
        });
        return result.error !== undefined;
    }
    catch (err) {
        return true;
    }
}
class RunScriptError extends Error {
    stdio;
    exitcode;
    constructor(message, stdio, exitcode, options) {
        super(message, options);
        this.name = this.constructor.name;
        this.stdio = stdio;
        this.exitcode = exitcode;
        Error.captureStackTrace(this, this.constructor);
    }
}
exports.RunScriptError = RunScriptError;
class RunScriptTimeoutError extends Error {
    stdio;
    timeout;
    constructor(message, stdio, timeout, options) {
        super(message, options);
        this.name = this.constructor.name;
        this.stdio = stdio;
        this.timeout = timeout;
        Error.captureStackTrace(this, this.constructor);
    }
}
exports.RunScriptTimeoutError = RunScriptTimeoutError;
/**
 * Run shell script in child process
 * Support OSX, Linux and Windows
 * @param {String} script - full script string, like `git clone https://github.com/node-modules/runscript.git`
 * @param {Object} [options] - spawn options
 *   @see https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options
 * @param {Object} [extraOptions] - extra options for running
 *   - {Number} [extraOptions.timeout] - child process running timeout
 * @return {Object} stdio object, will contains stdio.stdout and stdio.stderr buffer.
 */
function runScript(script, options = {}, extraOptions = {}) {
    return new Promise((resolve, reject) => {
        options.env = options.env || Object.assign({}, process.env);
        options.cwd = options.cwd || process.cwd();
        if (typeof options.cwd === 'object') {
            // convert URL object to string
            options.cwd = String(options.cwd);
        }
        options.stdio = options.stdio || 'inherit';
        if (options.stdout) {
            (0, node_assert_1.default)((0, is_type_of_1.isWritable)(options.stdout), 'options.stdout should be writable stream');
        }
        if (options.stderr) {
            (0, node_assert_1.default)((0, is_type_of_1.isWritable)(options.stderr), 'options.stderr should be writable stream');
        }
        let sh = 'sh';
        let shFlag = '-c';
        if (process.platform === 'win32') {
            sh = process.env.comspec || 'cmd';
            shFlag = '/d /s /c';
            options.windowsVerbatimArguments = true;
            if (script.indexOf('./') === 0 || script.indexOf('.\\') === 0 ||
                script.indexOf('../') === 0 || script.indexOf('..\\') === 0) {
                const splits = script.split(' ');
                // in bash C:\Windows\system32 -> C:\\Windows\\system32
                splits[0] = node_path_1.default.join(isCmd() ? options.cwd : node_path_1.default.normalize(options.cwd), splits[0]);
                script = splits.join(' ');
            }
        }
        debug('%s %s %s, %j, %j', sh, shFlag, script, options, extraOptions);
        const proc = (0, node_child_process_1.spawn)(sh, [shFlag, script], options);
        const stdout = [];
        const stderr = [];
        let isEnd = false;
        let timeoutTimer;
        if (proc.stdout) {
            proc.stdout.on('data', (buf) => {
                debug('stdout %d bytes', buf.length);
                stdout.push(buf);
            });
            if (options.stdout) {
                proc.stdout.pipe(options.stdout);
            }
        }
        if (proc.stderr) {
            proc.stderr.on('data', (buf) => {
                debug('stderr %d bytes', buf.length);
                stderr.push(buf);
            });
            if (options.stderr) {
                proc.stderr.pipe(options.stderr);
            }
        }
        proc.on('error', err => {
            debug('proc emit error: %s', err);
            if (isEnd) {
                return;
            }
            isEnd = true;
            clearTimeout(timeoutTimer);
            reject(err);
        });
        proc.on('exit', code => {
            debug('proc emit exit: %s', code);
            if (isEnd) {
                return;
            }
            isEnd = true;
            clearTimeout(timeoutTimer);
            const stdio = {
                stdout: null,
                stderr: null,
            };
            if (stdout.length > 0) {
                stdio.stdout = Buffer.concat(stdout);
            }
            if (stderr.length > 0) {
                stdio.stderr = Buffer.concat(stderr);
            }
            if (code !== 0) {
                const err = new RunScriptError(`Run "${sh} ${shFlag} ${script}" error, exit code ${code}`, stdio, code);
                return reject(err);
            }
            return resolve(stdio);
        });
        proc.on('close', code => {
            debug('proc emit close: %s', code);
        });
        if (typeof extraOptions.timeout === 'number' && extraOptions.timeout > 0) {
            const timeout = extraOptions.timeout;
            // start timer
            timeoutTimer = setTimeout(() => {
                debug('proc run timeout: %dms', timeout);
                isEnd = true;
                debug('kill child process %s', proc.pid);
                proc.kill();
                const stdio = {
                    stdout: null,
                    stderr: null,
                };
                if (stdout.length > 0) {
                    stdio.stdout = Buffer.concat(stdout);
                }
                if (stderr.length > 0) {
                    stdio.stderr = Buffer.concat(stderr);
                }
                const err = new RunScriptTimeoutError(`Run "${sh} ${shFlag} ${script}" timeout in ${extraOptions.timeout}ms`, stdio, timeout);
                return reject(err);
            }, timeout);
        }
    });
}
exports.runscript = runScript;
